core/vm/validation.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2021 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+)
+
+// validateCode checks that there're no undefined instructions and code ends with a terminating instruction
+func validateCode(code []byte, codeSections int, jumpTable *JumpTable) error {
+	var (
+		i        = 0
+		analysis []byte
+		opcode   OpCode
+	)
+	for i < len(code) {
+		switch opcode = OpCode(code[i]); {
+		case jumpTable[opcode].undefined:
+			return fmt.Errorf("%v: %v", ErrEOF1UndefinedInstruction, opcode)
+		case opcode >= PUSH1 && opcode <= PUSH32:
+			i += int(opcode) - int(PUSH1) + 2
+			continue // todo make sure this actually continues
+		case opcode == RJUMP || opcode == RJUMPI:
+			var arg int16
+			// Read immediate argument.
+			if err := binary.Read(bytes.NewReader(code[i+1:]), binary.BigEndian, &arg); err != nil {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Calculate relative target.
+			pos := i + 3 + int(arg)
+
+			// Check if offset points to out-of-bounds location.
+			if pos < 0 || pos >= len(code) {
+				return ErrEOF1InvalidRelativeOffset
+			}
+			// Check if offset points to non-code segment.
+			// TODO(matt): include CALLF and RJUMPs in analysis.
+			if analysis == nil {
+				// TODO CZ: implement the logic https://github.com/ethereum/go-ethereum/pull/26133/commits/370c24c344f7fc26cc29fba9b33825baeded8876#diff-882af11429522494f308c57017fe1c44e86b338e3bf5c308584d9d187606d252L64
```

Comment:
This has to be implemented

Reply:
resolved https://github.com/ETCCooperative/erigon/pull/4/commits/3b35f1d5602b9aaac5b702c486d9e7b64a1a92b6